Analyzer to detect RequiresPreviewFeatures attribute

pgovind opened the pull request:
Partly addresses dotnet/runtime#52726

I think this is a good point to get the implementation reviewed. I've borrowed ideas from the PlatformCompatilityAnalyzer implementation.

Pending work (in a follow up PR):

1. Update unit tests to include VB once all the C# test cases have been defined comprehensively
2. Another PR to start consuming the latest sdk in this repo so that we can do symbol level comparison of `RequiresPreviewFeaturesAttribute` and also implement logic to look for the attribute on dependencies.
3 Replace UseFirstDescriptor with the DiagnosticID for each test case
4. Fix the TODO comments in the analyzer

pgovind commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,320 @@
+        private static readonly ImmutableArray<SymbolKind> s_symbols = ImmutableArray.Create(SymbolKind.NamedType, SymbolKind.Method, SymbolKind.Property, SymbolKind.Field, SymbolKind.Event);
+
+#pragma warning disable RS0030 // The symbol 'DiagnosticDescriptor.DiagnosticDescriptor.#ctor' is banned in this project: Use 'DiagnosticDescriptorHelper.Create' instead
+        internal static DiagnosticDescriptor GeneralPreviewFeatureAttributeRule = new(RuleId,
```
DiagnosticDescriptionHelper.Create doesn't expose a way to specify `DiagnosticSeverity.Error`. My thought here is that `Error` is a very rare occurence, so we shouldn't expose it from `DiagnosticDescriptionHelper.Create`. So I'm using the constructor directly instead. Let me know if you want me to  change it

Youssef1313 replied:
@pgovind The helper method adds some custom tags which I believe are used for generating some globalconfigs. It's likely that using the ctor on the current form (without custom tags) will be causing a bug when setting AnalysisMode or something. I personally think it's okay to expose a new member in RuleLevel enum and use the helper, even if error is rare. But @mavasani can confirm what's the best action here.

mavasani replied:
Yes, @Youssef1313. It is instead better to add a new RuleLevel for BuildError: https://github.com/dotnet/roslyn-analyzers/blob/main/src/Utilities/Compiler/RuleLevel.cs#L9.

pgovind commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,320 @@
+                IAssemblySymbol? thisAssembly = compilation.Assembly;
+                ImmutableArray<AttributeData> attributes = thisAssembly.GetAttributes();
+
+                AttributeData? attribute = attributes.FirstOrDefault(x => x.AttributeClass.Name == RequiresPreviewFeaturesAttribute);
```
I can't do a symbol comparison here yet (already called this out in the PR description). The RequiresPreviewFeaturesAttribute was introduced in .NET6P5, and this project doesn't use that version yet. I'd like to fix this code to use symbol comparison in a later PR after:
1. The P6 sdk becomes public
2. I update the 6.0.0.x branch to use the P6 sdk

Youssef1313 replied:
@pgovind I think symbol comparison can be done without .NET 6 P5.

If the blocker is unit tests, then the attribute in tests just needs to be defined in the correct namespace.

Youssef1313 replied:
Since this is used only for null check. This could use `Any` instead, or it can use HasAttribute whenever symbol comparison is used.

buyaa-n replied:
Sounds better to add mock attribute class for the unit testing and do symbol comparison here, it could fail fast if the new attribute is not found

pgovind replied:
Fixed now

pgovind commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,320 @@
+
+        private static bool PropertyOrMethodSymbolHasAttribute(ISymbol propertyOrMethodSymbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, ConcurrentDictionary<ISymbol, PreviewFeatureUsageType> requiresPreviewFeaturesSymbolsToUsageType, INamedTypeSymbol? runtimeFeatureType)
+        {
+            if (SymbolIsStaticAndAbstract(propertyOrMethodSymbol, requiresPreviewFeaturesSymbols))
```
Unfortunately we can't test this yet. `public abstract static void Foo()` will not compile with the current version of the compiler. Once the new compiler is consumable in an sdk build, I will update the branch with that build and add new unit tests

buyaa-n replied:
Did you try to test by setting the `LanguageVersion`? https://github.com/dotnet/roslyn-analyzers/blob/a0522afd52b25ad211298e308fd5bbc7734865e0/src/Roslyn.Diagnostics.Analyzers/UnitTests/DoNotCopyValueTests.cs#L25
Not sure if the version you need available there, though it has even preview option https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/LanguageVersion.cs,198

pgovind replied:
> Did you try to test by setting the LanguageVersion

I don't think that will work yet. To use `abstract static` we will need to start consuming the latest compiler from the latest sdk itself. Once this PR goes in, I will update this branch to consume the latest daily build of the sdk and then I can add a new unit test for this code path.

pgovind commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,320 @@
+                }
+            }
+
+            if (propertyOrMethodSymbol.IsImplementationOfAnyImplicitInterfaceMember(out ISymbol baseInterfaceMember))
```
This is a new extension method. The existing method doesn't expose `baseInterfaceMember`. Instead it only returned a bool.

pgovind commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureUnitTests.cs:
```diff
@@ -0,0 +1,1060 @@
+using VerifyCS = Test.Utilities.CSharpCodeFixVerifier<
+    Microsoft.NetCore.Analyzers.Runtime.DetectPreviewFeatureAnalyzer,
+    Microsoft.CodeAnalysis.Testing.EmptyCodeFixProvider>;
+using VerifyVB = Test.Utilities.VisualBasicCodeFixVerifier<
```
Unused. Remove

pgovind commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureUnitTests.cs:
```diff
@@ -0,0 +1,1060 @@
+                        csInput
+                    }
+                },
+                MarkupOptions = MarkupOptions.UseFirstDescriptor,
```
I figured this out from running the unit tests with multiple diagnostics. Just out of curiosity, what is the right way to specify which diagnostic to test for with the markup syntax?

Youssef1313 replied:
@pgovind I believe it's `{|ID:code|}`

pgovind commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureUnitTests.cs:
```diff
@@ -0,0 +1,1060 @@
+
+        }
+
+        // TODO: The following test cannot be activated yet because it requires preview roslyn bits
```
Let me know if you want this commented out portion removed.

pgovind commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureUnitTests.cs:
```diff
@@ -0,0 +1,1060 @@
+    [RequiresPreviewFeatures]
+    void CallSite()
+    {
+        PreviewMethod();
```
This test will fail right now because I don't know what the right way to detect this is. If I inspect `IOperation` here, it will of type `IInvocation`. My plan was to go up the parent chain here to inspect the `CallSite` method declaration and check if the `reqeuiresPreviewFeatures` attribute had been set. However I'm running into problems. The parent chain here (starting from `PreviewMethod()`) seems to be `ExpressionOperation -> BlockOperation -> MethodBodyOperation`. `MethodBodyOperation`'s parent in `null`, so I'm unable to get to the `CallSite`'s definition. However, if I inspect the `Syntax` property of `MethodBodyDeclaration` and traverse it's parent, I can get to `CallSite`'s syntax node. This would be C# specific however, so it is not an option. Is there another API that let's me inspect `CallSite` properly?

@mavasani @buyaa-n

pgovind replied:
Just calling this comment out again. I'm wondering if you guys know the answer or a better way here? @mavasani @buyaa-n

buyaa-n replied:
> My plan was to go up the parent chain here to inspect the CallSite method declaration and check if the reqeuiresPreviewFeatures attribute had been set. However I'm running into problems. The parent chain here (starting from PreviewMethod()) seems to be ExpressionOperation -> BlockOperation -> MethodBodyOperation. MethodBodyOperation's parent in null, so I'm unable to get to the CallSite's definition.

You should check the ContainingSymbol, not the parent, see how it is done in PCA:
https://github.com/dotnet/roslyn-analyzers/blob/582230ea924cba9f2e2240b852b12048c07aa31c/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs#L1050-L1070 
Here on row 1050 checks if the operation has platform attribute, if it has then loading the containing symbol's (i. e callsite) attributes on row 1058, then on row 1061 checking if the call site attributes can suppress the operation attributes

Note that here the `TryGetOrCreatePlatformAttributes` has a recursive call and collects all attributes from symbol to assembly, not only go up one level

pgovind replied:
> You should check the ContainingSymbol, not the parent, see how it is done in PCA

This is what the code is doing now, but it doesn't work I think. For this test case:

``` csharp
public class Program
{
    [RequiresPreviewFeatures]
    public virtual void PreviewMethod()  { }
   
    [RequiresPreviewFeatures]
    void CallSite()
    {
        PreviewMethod();
    }
}
}
```

`PreviewMethod()` (from inside CallSite) will be accessed as an `INamedTypeSymbol`. It's `ContainingType/Symbol` will be `Program`, not `CallSite`, because `Program` is the one that defines `PreviewMethod`. What I need is a way to know that the current `PreviewMethod` invocation has been called from within another method that is also marked as Preview. I wonder if I should be using SemanticModel here. I'll explore locally

buyaa-n replied:
> PreviewMethod() (from inside CallSite) will be accessed as an INamedTypeSymbol. It's ContainingType/Symbol will be Program, not CallSite, because Program is the one that defines PreviewMethod.

That doesn't sound right, the `ContainingType` of the PreviewMethod() invocation is the `CallSite()` method, not `Program`, `Program` is the  `ContainingType` of the `PreviewMethod` declaration

buyaa-n replied:
`PreviewMethod()` invocation should be flagged from operation analysis not from symbol analysis, probably there is duplication in your analysis, anyway, I believe the diagnostic is being reported in the `PreviewMethod()` invocation not in the declaration, so from there you should be able to get the real `CallSite()` using `ContainingType`

pgovind replied:
Calling context.ContainingSymbol worked. This is now resolved.

pgovind commented:
I think this is ready to review now.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1611,4 +1611,19 @@
+    <value>An assembly has to opt into preview features before using them.</value>
+  </data>
+  <data name="DetectPreviewFeaturesMessage" xml:space="preserve">
+    <value>Using {0} requires opting into preview features. See https://aka.ms/dotnet-warnings/preview-features for more information.</value>
```
Placeholders are usually put between single quotes

```suggestion
    <value>Using '{0}' requires opting into preview features. See https://aka.ms/dotnet-warnings/preview-features for more information.</value>
```

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1611,4 +1611,19 @@
+    <value>This API requires opting into preview features</value>
+  </data>
+  <data name="ImplementsEmptyPreviewInterface" xml:space="preserve">
+    <value>{0} implements a preview interface.  See https://aka.ms/dotnet-warnings/preview-features for more information.</value>
```
- Placeholders are usually put between single quotes.
- Also removed a seemingly extra space

```suggestion
    <value>'{0}' implements a preview interface. See https://aka.ms/dotnet-warnings/preview-features for more information.</value>
```

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,320 @@
+                {
+                    if (usageType == PreviewFeatureUsageType.StaticAbstract)
+                    {
+                        context.ReportDiagnostic(symbol.CreateDiagnostic(StaticAbstractIsPreviewFeatureRule));
```
Will that get reported for something like:

```csharp
public abstract class C
{
    public static abstract M();
}
```

I think it shouldn't, since the above is an error anyway (static abstract is only allowed in interfaces), and this diagnostic will make it feel like this is allowed in preview.

pgovind replied:
It will yea. However, I am unable to test it at the moment because we don't consume the latest compiler yet. Without testing, I'm not sure how the parent chain will look here. I've added a TODO for now. Once we started consuming the latest compiler here, I can fix this.

mavasani replied:
Can you file a tracking issue for this and assign it to yourself?

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,320 @@
+                    operation = catchClauseOperation.ExceptionDeclarationOrExpression;
+                }
+
+                context.ReportDiagnostic(operation.CreateDiagnostic(GeneralPreviewFeatureAttributeRule, symbol!.Name));
```
Instead of the suppression, shouldn't adding `[NotNullWhen(true)]` to the out parameter of `OperationUsesPreviewFeatures` fix that?

pgovind replied:
Nice catch. Fixed.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,320 @@
+            {
+                IInvocationOperation iOperation => iOperation.TargetMethod,
+                IObjectCreationOperation cOperation => cOperation.Constructor,
+                IPropertyReferenceOperation pOperation => pOperation.Property,
```
Are parameter reference and local reference operations skipped intentionally?

pgovind replied:
Yup. My reasoning was:
If a preview type is passed in as a parameter (or is a local variable or is a type instance), the use of any method/property/field on that parameter/variable/type will be marked by the analyzer. Or the variable remains unused. I can't think of a common scenario where we miss tagging preview feature usage.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,320 @@
+                IEventReferenceOperation eOperation => eOperation.Member,
+                IUnaryOperation uOperation => uOperation.OperatorMethod,
+                IBinaryOperation bOperation => bOperation.OperatorMethod,
+                IArrayCreationOperation arrayCreationOperation => arrayCreationOperation.Type is IArrayTypeSymbol arrayTypeSymbol ? arrayTypeSymbol.ElementType : null,
```
```suggestion
                IArrayCreationOperation arrayCreationOperation => (arrayCreationOperation.Type as IArrayTypeSymbol)?.ElementType,
```

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,320 @@
+                                                                                      DiagnosticSeverity.Error,
+                                                                                      isEnabledByDefault: true,
+                                                                                      s_localizableDescription,
+                                                                                      helpLinkUri: @" https://aka.ms/dotnet-warnings/preview-features");
```
there is an extra space here + links should point to docs.microsoft.com.

Youssef1313 commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureUnitTests.cs:
```diff
@@ -0,0 +1,1019 @@
+namespace Preview_Feature_Scratch
+{" + setupPreviewAttribute +
+@"
+    public class Program
+    {
+        static void Main(string[] args)
+        {
+            var a = new Fraction();
+            var b = [|+a|];
+        }
+    }
+
+    public readonly struct Fraction
+    {
+        [RequiresPreviewFeatures]
+        public static Fraction operator +(Fraction a) => a;
+    }
+}
```
What should happen if `Fraction` is defined in a generated file, but `Program` is not? Should there be a diagnostic?

i.e, should we use `GeneratedCodeAnalysisFlags.Analyze`?

pgovind replied:
Hmm interesting. I hadn't seen that enum before. I think having `Analyze` makes sense. Have you seen a unit test that tests this behavior?

Youssef1313 replied:
@pgovind Do you mean testing generated code? I think you need to pass two sources, and add `// <auto-generated/>` comment on one of them.

buyaa-n replied:
> Have you seen a unit test that tests this behavior

I don't think we need to test if the generated code analysis working

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,320 @@
+
+                if (context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeCompilerServicesRuntimeFeature, out INamedTypeSymbol? runtimeFeatureType))
+                {
+                    if (runtimeFeatureType != null)
```
Redundant check, out parameter has `[NotNullWhen(returnValue: true)]` 
```suggestion
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,320 @@
+                {
+                    if (runtimeFeatureType != null)
+                    {
+                        HasPreviewAttribute(runtimeFeatureType, requiresPreviewFeaturesSymbols);
```
it's a bit strange, need to check the type loaded from the runtime?  not existing in the user code?

pgovind replied:
Yup, because static abstract is considered preview if runtimeFeatureType has the Preview attribute on it. So, in a future release, if runtimeFeatureType is not marked as preview, we won't consider static abstract as preview.

pgovind replied:
Fixed now. I'm looking for the attribute on RuntimeFeature.VirtualStaticsInInterfaces now

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,320 @@
+        private static bool OperationUsesPreviewFeatures(IOperation operation, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, out ISymbol? symbol)
+        {
+            symbol = GetOperationSymbol(operation);
+            // CHange the operation here to ExceptionDeclarationOrExpression
```
Outdated comment or?

AraHaan commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1611,4 +1611,19 @@
+    <value>{0} implements a preview interface.  See https://aka.ms/dotnet-warnings/preview-features for more information.</value>
+  </data>
+  <data name="StaticAndAbstractRequiresPreviewFeatures" xml:space="preserve">
+    <value>Using both static and abstract modifiers requires opting into preview features. See https://aka.ms/dotnet-warnings/preview-features for more information.</value>
```
Link does not seem to work.

pgovind replied:
It's actually not up yet :) I expect it will be up after Preview 7 drops and/or when the docs team does their sweep of the repo for new docs to be created

AraHaan replied:
I expected there to be an docs list of currently available preview features atm.

Youssef1313 commented on src/Utilities/Compiler/DiagnosticDescriptorHelper.cs:
```diff
@@ -60,6 +60,7 @@ internal static partial class DiagnosticDescriptorHelper
                     RuleLevel.IdeHidden_BulkConfigurable => (DiagnosticSeverity.Hidden, true),
                     RuleLevel.Disabled => (DiagnosticSeverity.Warning, false),
                     RuleLevel.CandidateForRemoval => (DiagnosticSeverity.Warning, false),
+                    RuleLevel.BuildError => (DiagnosticSeverity.Error, true),
```
`DiagnosticDescriptorCreationAnalyzer` may need to get updated to handle this new RuleLevel.

pgovind replied:
It could be updated yea, but we should do it in a separate PR once this goes in

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,321 @@
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
```
Analyzing generated code without reporting the diagnostics doesn't make sense
```suggestion
            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
```

Youssef1313 replied:
@buyaa-n I believe it does. We want to collect data about generated code, but only report for non-generated code.

buyaa-n replied:
Probably there is a case where it makes sense, especially when the diagnostic is not important, for this analyzer i believe it is important to report diagnostics, but i will let @pgovind decide

Youssef1313 replied:
@buyaa-n I thought it was agreed on that only security analyzers should report for auto-generated code?

buyaa-n replied:
Oh yeah? Did not know what, do you have a reference to that agreement? I would like to see that if there is a reference. 

By accounting, this analyzer is error level and it will be quite expensive i would expect it to have either `GeneratedCodeAnalysisFlags.None` or if we need to analyze generated code it should report the diagnostics too, but i will let @pgovind  and repo owners decide CC @jmarolf @sharwell

stephentoub replied:
We should cover generated code as well.  Any use of a preview feature when not in preview, whether that use comes from handwritten or generated code, should be flagged.

Youssef1313 replied:
> do you have a reference to that agreement?

I don't recall where the comment was. But since this is the "first" analyzer with error severity, it may be an exception that it reports on generated code as well. cc @mavasani

mavasani replied:
> But since this is the "first" analyzer with error severity, it may be an exception that it reports on generated code as well

That sounds a valid reason for this code to report diagnostics in generated code.

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,321 @@
+                ImmutableArray<AttributeData> attributes = thisAssembly.GetAttributes();
+
+                AttributeData? attribute = attributes.FirstOrDefault(x => x.AttributeClass == previewFeaturesAttribute);
+                if (!(attribute is null))
```
```suggestion
                if (attribute is not null)
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,321 @@
+                    return;
+                }
+
+                Compilation compilation = context.Compilation;
+                IAssemblySymbol? thisAssembly = compilation.Assembly;
+                ImmutableArray<AttributeData> attributes = thisAssembly.GetAttributes();
+
+                AttributeData? attribute = attributes.FirstOrDefault(x => x.AttributeClass == previewFeaturesAttribute);
```
NIT: IMO these variables are redundant as not used anywhere else, but if you prefer this way it's OK, up to you
```suggestion
                AttributeData? attribute = context.Compilation.Assembly.GetAttributes().FirstOrDefault(x => x.AttributeClass == previewFeaturesAttribute);
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,321 @@
+
+            if (containingSymbol != null)
+            {
+                if (HasPreviewAttribute(containingSymbol, requiresPreviewFeaturesSymbols))
```
Is it enough to check only one level up? If the containing symbol/method was not annotated but what if the contacting type or assembly was annotated?

pgovind replied:
> assembly was annotated

If the current assembly is annotated, we will detect it. If dependencies are annotated, we won't currently. I'll implement that in another PR (also called it out in the PR description). The aim here is to get this PR in soon so Tanner can give me more feedback. 

> Is it enough to check only one level up?

For now this seems to work fine. I haven't come across a case where we need to look 2 levels up. If we do, then we can change it. Looking only 1 level up feels clean. Any more than that and I fear we will have holes in our detection.

buyaa-n replied:
> If the current assembly is annotated, we will detect it. If dependencies are annotated, we won't currently. 

I meant what if the assembly of referenced API is annotated, not the current assembly

> The aim here is to get this PR in soon so Tanner can give me more feedback.

You can also give him the DLL of your analyzer and he can replace the analyzer wherever it is using

pgovind replied:
> I meant what if the assembly of referenced API is annotated

Right right, that's what I meant. We don't detect it right now. That is by design because I want to get this PR in first. I'll add that detection in a follow up PR.

> You can also give him the DLL of your analyzer and he can replace the analyzer wherever it is using

We tried that 2-3 weeks back and it didn't work unfortunately. We also couldn't figure out why. I think it had something to do with using the latest compiler + sdk + analyzer bits combination :/ That's why we decided to get the first pass of this analyzer in, so that we can get some feedback while Tanner writes unit tests

buyaa-n replied:
👌

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,321 @@
+
+        private static bool OperationUsesPreviewFeatures(ISymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols)
+        {
+            var containingSymbol = symbol.ContainingSymbol;
+            // Namespaces do not have attributes
+            while (containingSymbol is INamespaceSymbol)
+            {
+                containingSymbol = containingSymbol.ContainingSymbol;
+            }
+
+            if (containingSymbol != null)
+            {
+                if (HasPreviewAttribute(containingSymbol, requiresPreviewFeaturesSymbols))
```
Please move the logic to the `ContainingTypeIsPreview` method  https://github.com/dotnet/roslyn-analyzers/blob/f147d906818271152d87beac60a6ab3c6808772b/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs#L138 and use the method here

pgovind replied:
That method doesn't special case INamespaceSymbol though. I can do it if you think it's important, but I'd prefer to keep it this way.

buyaa-n replied:
> That method doesn't special case INamespaceSymbol though.

That is why I said move this logic, i do not understand why one of the same logic special case it but the other not, if you have a reason for that then its OK, you can keep it as is

pgovind replied:
As it turns out the other method will never see a NamespaceSymbol, so it was safe to move this logic there.

buyaa-n commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureUnitTests.cs:
```diff
@@ -0,0 +1,1021 @@
+
+            await new VerifyCS.Test
+            {
+                LanguageVersion = CodeAnalysis.CSharp.LanguageVersion.CSharp9,
+                TestState =
+                {
+                    Sources =
+                    {
+                        csInput, requiresPreviewFeaturesAttribute
+                    }
+                },
+                MarkupOptions = MarkupOptions.UseFirstDescriptor,
```
SEems used for all/many tests, please use a static helper function

buyaa-n commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureUnitTests.cs:
```diff
@@ -0,0 +1,1021 @@
+    }
+
+    [RequiresPreviewFeatures]
+    public class Lib
```
Is this class used?

pgovind replied:
Nope. Removed now.

buyaa-n commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureUnitTests.cs:
```diff
@@ -0,0 +1,1021 @@
+                    }
+                },
+                MarkupOptions = MarkupOptions.UseFirstDescriptor,
+            }.RunAsync();
```
Probably check the Diagnostic and/or message for some of the tests as they have different Diagnostics on different scenarios

pgovind replied:
Yup, that is on my todo. Do you mind if I do it in a follow up PR?

buyaa-n replied:
🆗

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,321 @@
+            return HasPreviewAttribute(symbol, requiresPreviewFeaturesSymbols);
+        }
+
+        private static bool HasPreviewAttribute(ISymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols)
```
`HasXXX` generally indicates that the method has no side effects, but that doesn't seem to be the case here as the passed in dictionary can be modified. Consider renaming to `TryProcessPreviewAttribute` or `ProcessPreviewAttribute`

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,321 @@
+            if (!requiresPreviewFeaturesSymbols.TryGetValue(symbol, out bool existing))
+            {
+                bool ret = false;
+                ImmutableArray<AttributeData> attributes = symbol.GetAttributes();
+                foreach (AttributeData attribute in attributes)
+                {
+                    string attributeName = attribute.AttributeClass.Name;
+                    if (attributeName == RequiresPreviewFeaturesAttribute)
+                    {
+                        ret = true;
+                        break;
+                    }
+                }
```
Can you instead use this extension method? https://github.com/dotnet/roslyn-analyzers/blob/a0522afd52b25ad211298e308fd5bbc7734865e0/src/Utilities/Compiler/Extensions/ISymbolExtensions.cs#L595

pgovind replied:
Added a TODO. Can't do this yet, because we need to consume the latest sdk to get the `RequiresPreviewFeature` attribute symbol.

Youssef1313 replied:
> Can't do this yet, because we need to consume the latest sdk to get the `RequiresPreviewFeature` attribute symbol.

@pgovind I believe this is outdated now?

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,321 @@
+                IAssemblySymbol? thisAssembly = compilation.Assembly;
+                ImmutableArray<AttributeData> attributes = thisAssembly.GetAttributes();
+
+                AttributeData? attribute = attributes.FirstOrDefault(x => x.AttributeClass == previewFeaturesAttribute);
+                if (!(attribute is null))
+                {
+                    // This assembly has enabled preview attributes.
+                    return;
+                }
```
```suggestion
                if (thisAssembly.HasAttribute(previewFeaturesAttribute))
                {
                    // This assembly has enabled preview attributes.
                    return;
                }
```

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,321 @@
+                    return;
+                }
+
+                ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols = new ConcurrentDictionary<ISymbol, bool>();
+                ConcurrentDictionary<ISymbol, PreviewFeatureUsageType> requiresPreviewFeaturesSymbolsToUsageType = new ConcurrentDictionary<ISymbol, PreviewFeatureUsageType>();
```
Nit:
```suggestion
                ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols = new();
                ConcurrentDictionary<ISymbol, PreviewFeatureUsageType> requiresPreviewFeaturesSymbolsToUsageType = new();
```

mavasani replied:
Or use `var x = new XYZ();`, whichever style you prefer.

AraHaan commented:
Also not sure if anyone else noticed but ``RequiresPreviewFeatures`` is spelled wrong on the PR's title.

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,321 @@
+        private static void AnalyzeSymbol(SymbolAnalysisContext context, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, ConcurrentDictionary<ISymbol, PreviewFeatureUsageType> requiresPreviewFeaturesSymbolsToUsageType, INamedTypeSymbol? runtimeFeatureType)
+        {
+            ISymbol symbol = context.Symbol;
+            if (symbol == null)
+            {
+                return;
+            }
```
This is not possible
```suggestion
```

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,321 @@
+            bool reportDiagnostic = false;
+
+            // Look for attributes on base classes that this type may extend
+            if (symbol is ITypeSymbol typeSymbol)
+            {
+                reportDiagnostic = TypeSymbolHasAttribute(typeSymbol, requiresPreviewFeaturesSymbols, requiresPreviewFeaturesSymbolsToUsageType);
+            }
+
+            if (!reportDiagnostic)
+            {
+                if (symbol is IMethodSymbol || symbol is IPropertySymbol)
+                {
+                    reportDiagnostic = PropertyOrMethodSymbolHasAttribute(symbol, requiresPreviewFeaturesSymbols, requiresPreviewFeaturesSymbolsToUsageType, runtimeFeatureType);
+                }
+            }
```
This would be more readable as a switch expression:

```c#
var reportDiagnostic = symbol switch
{
   ITypeSymbol typeSymbol => TypeSymbolHasAttribute(typeSymbol, requiresPreviewFeaturesSymbols, requiresPreviewFeaturesSymbolsToUsageType);
   IPropertySymbol _ or IMethodSymbol _ => PropertyOrMethodSymbolHasAttribute(symbol, requiresPreviewFeaturesSymbols, requiresPreviewFeaturesSymbolsToUsageType, runtimeFeatureType);
   _ => false;
}

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,321 @@
+        private static bool OperationUsesPreviewFeatures(IOperation operation, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, [NotNullWhen(true)] out ISymbol? symbol)
+        {
+            symbol = GetOperationSymbol(operation);
+            if (symbol == null)
+            {
+                return false;
+            }
+
+            return OperationUsesPreviewFeatures(symbol, requiresPreviewFeaturesSymbols);
```
Nit:
```suggestion
            return symbol != null && OperationUsesPreviewFeatures(symbol, requiresPreviewFeaturesSymbols);
```

AraHaan commented:
Can somone please rebase and resolve the merge conflicts?

pgovind commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,306 @@
+        {
+            IOperation operation = context.Operation;
+            symbol = GetOperationSymbol(operation);
+            if (operation is IInvocationOperation || operation is IPropertyReferenceOperation)
```
Let me know if someone prefers the switch expression here. Since it's only 2 types here, I didn't bother with `switch`

buyaa-n replied:
I wonder why it should apply only these 2, seems could apply for all operations

pgovind replied:
Good point. When I wrote this first I wanted to narrow the scope, but I don't see why it shouldn't be all operations. Fixed it now by removing the `if` condition. This should now also work if say, a preview operator + was used inside another preview method.

buyaa-n replied:
> Fixed it now by removing the if condition.

Do not see it is removed 😆

pgovind replied:
Definitely removed now

pgovind commented:
Alright I think this is ready for review again. I fixed the failing unit tests and addressed all the remaining comments. Should be good to go.

@buyaa-n @mavasani 

> Can someone please rebase and resolve the merge conflicts?
Done.

> Also not sure if anyone else noticed but RequiresPreviewFeatures is spelled wrong on the PR's title.
Ah yes. I've typed it so many times at this point that 1 of them was bound to be misspelled :D

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,307 @@
+            if (operation is IInvocationOperation || operation is IPropertyReferenceOperation)
+            {
+                ISymbol containingSymbol = context.ContainingSymbol;
+                if (containingSymbol is IMethodSymbol || containingSymbol is IPropertySymbol)
+                {
+                    if (OperationUsesPreviewFeatures(containingSymbol, requiresPreviewFeaturesSymbols) || OperationUsesPreviewFeatures(containingSymbol.ContainingSymbol, requiresPreviewFeaturesSymbols))
```
Same question here, it should apply for all symbols as attribute could be applied for all of them, i guess you are planned to cover that later

pgovind replied:
Fixed this now with the preview comment

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,307 @@
+            ISymbol symbol = context.Symbol;
+
+            // Don't report diagnostics on types that are marked as Preview
+            if (ProcessPreviewAttribute(symbol, requiresPreviewFeaturesSymbols))
```
Wondering what makes sure that it was a type symbol, seems this analyses many different symbols `SymbolKind.NamedType, SymbolKind.Method, SymbolKind.Property, SymbolKind.Field, SymbolKind.Event`

pgovind replied:
This was just a bad comment :) I meant everything when I wrote "types" here. Deleted the comment now. For ex: It's ok to see the `IMethodSymbol` for a preview method here. We will just add the symbol to the dictionary and return without reporting a diagnostic.

buyaa-n commented:
> Alright I think this is ready for review again. I fixed the failing unit tests and addressed all the remaining comments. Should be good to go.

@pgovind not sure if you notice, but the build failed

> Partly addresses dotnet/runtime#52726

This analyzer is Error level, so this can be breaking, especially by accounting that the PR partially covers the issue requirements I suggest decreasing the severity into `Info` level for now (you can set up it to an error level manually in runtime or on any project). We can update the severity to Error after all requirements are covered. What do you think about that @pgovind?

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,304 @@
+                                                                                                                    s_localizableTitle,
+                                                                                                                    s_localizableMessage,
+                                                                                                                    DiagnosticCategory.Usage,
+                                                                                                                    RuleLevel.BuildWarning,
```
```suggestion
                                                                                                                    RuleLevel.IdeSuggestion,
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,304 @@
+                                                                                                                    s_localizableTitle,
+                                                                                                                    s_staticAbstractMessage,
+                                                                                                                    DiagnosticCategory.Usage,
+                                                                                                                    RuleLevel.BuildWarning,
```
```suggestion
                                                                                                                    RuleLevel.IdeSuggestion,
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,304 @@
+                                                                                                                    s_localizableTitle,
+                                                                                                                    s_implementsEmptyPreviewInterface,
+                                                                                                                    DiagnosticCategory.Usage,
+                                                                                                                    RuleLevel.BuildWarning,
```
```suggestion
                                                                                                                    RuleLevel.IdeSuggestion,
```

pgovind replied:
Yup, fixed now

buyaa-n reviewed (changes requested):
Even `RuleLevel.BuildWarning` is breaking (warning) please change to `RuleLevel.IdeSuggestion` (info)

pgovind commented:
> What do you think about that @pgovind?

I'm ok with that. I decreased the severity to BuildWarning for now. I'll update it to Error in a follow up PR. I didn't remove the changes I made to add RuleLevel.Error though. I think it's ok since I'll be changing the severity to Error very soon. Let me know if someone wants me to remove them.

buyaa-n reviewed (approved):
One more comment https://github.com/dotnet/roslyn-analyzers/pull/5155/files#r666600012, other than that LGTM

mavasani commented on src/Utilities/Compiler/RuleLevel.cs:
```diff
@@ -47,6 +47,12 @@ internal enum RuleLevel
+        /// Correctness rule which should have <b>no false positives</b>, and is required to be fixed by users.
+        /// This rule will be <b>enabled in CI and IDE live analysis</b> by default with severity <see cref="DiagnosticSeverity.Error"/>.
+        /// </summary>
+        BuildError = 6,
```
I think it is better to give this value 1 and increment the rest of the fields, the enum goes from highest to lowest severity and its not a public API so a breaking change is fine.

mavasani commented on src/Utilities/Compiler/Extensions/ISymbolExtensions.cs:
```diff
@@ -466,6 +466,31 @@ public static bool IsImplementationOfAnyImplicitInterfaceMember<TSymbol>(this IS
+        /// <summary>
+        /// Checks if a given symbol implements an interface member implicitly
+        /// </summary>
+        public static bool IsImplementationOfAnyImplicitInterfaceMember<TSymbol>(this ISymbol symbol, out TSymbol interfaceMember)
```
Is this a method move or a new method? Note that the implementation here is pretty expensive as we are iterating all interface methods in the type hierarchy and invoking `IsImplementationOfInterfaceMember` which does expensive work. If this extension method is called from a hot path, you certainly want to do performance measurements/profiling for this analyzer.

mavasani replied:
Tagging @sharwell as he may have insights on the performance characteristics of this code.

pgovind replied:
Yup. Called it out here: https://github.com/dotnet/roslyn-analyzers/pull/5155#discussion_r654071422

> you certainly want to do performance measurements/profiling for this analyzer

👍

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,304 @@
+            });
+        }
+
+        private static bool TypeSymbolHasAttribute(ITypeSymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, ConcurrentDictionary<ISymbol, PreviewFeatureUsageType> requiresPreviewFeaturesSymbolsToUsageType)
```
Nit: Might be better to rename to `ProcessTypeSymbolAttributes`?

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,304 @@
+            return false;
+        }
+
+        private static bool ContainingTypeIsPreview(ISymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols)
```
Rename to `ProcessContainingTypePreviewAttribute`?

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,304 @@
+            {
+                containingType = containingType.ContainingType;
+            }
+            if (containingType != null)
```
Nit: New line between compound statements.
```suggestion

            if (containingType != null)
```

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,304 @@
+        }
+
+        private static bool SymbolIsStaticAndAbstract(ISymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols)
+        {
+            bool isStaticAndAbstract = symbol.IsStatic && symbol.IsAbstract;
+
+            return isStaticAndAbstract && !ContainingTypeIsPreview(symbol, requiresPreviewFeaturesSymbols);
+        }
```
Nit: Do we need the local?
```suggestion
            => symbol.IsStatic && symbol.IsAbstract && !ContainingTypeIsPreview(symbol, requiresPreviewFeaturesSymbols);
```

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,304 @@
+            return isStaticAndAbstract && !ContainingTypeIsPreview(symbol, requiresPreviewFeaturesSymbols);
+        }
+
+        private static bool PropertyOrMethodSymbolHasAttribute(ISymbol propertyOrMethodSymbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, ConcurrentDictionary<ISymbol, PreviewFeatureUsageType> requiresPreviewFeaturesSymbolsToUsageType, INamedTypeSymbol? runtimeFeatureType)
```
Rename to `ProcessPropertyOrMethodSymbolAttributes`?

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,304 @@
+            };
+
+        private static bool OperationUsesPreviewFeatures(ISymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols)
+        {
+            if (ContainingTypeIsPreview(symbol, requiresPreviewFeaturesSymbols))
+            {
+                return true;
+            }
+
+            return ProcessPreviewAttribute(symbol, requiresPreviewFeaturesSymbols);
+        }
```
Nit:
```suggestion
            => ContainingTypeIsPreview(symbol, requiresPreviewFeaturesSymbols)) ||
               ProcessPreviewAttribute(symbol, requiresPreviewFeaturesSymbols);
```

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,304 @@
+            if (!requiresPreviewFeaturesSymbols.TryGetValue(symbol, out bool existing))
+            {
+                bool ret = false;
+                // TODO: Use symbol.HasAttribute(RequiresPreviewFeaturesAttribute) once we consume the latest sdk
```
Can you file a tracking bug with all the TODOs and assigned it to yourself?

pgovind replied:
https://github.com/dotnet/roslyn-analyzers/issues/5227

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,304 @@
+            if (operation is IInvocationOperation || operation is IPropertyReferenceOperation)
+            {
+                ISymbol containingSymbol = context.ContainingSymbol;
+                if (OperationUsesPreviewFeatures(containingSymbol, requiresPreviewFeaturesSymbols) || OperationUsesPreviewFeatures(containingSymbol.ContainingSymbol, requiresPreviewFeaturesSymbols))
```
Any reason we are only looking up one level of `.ContainingSymbol` and not walking the entire `.ContainingSymbol` chain?

pgovind replied:
No reason. The symbol analysis stops checking after 1 level, so I decided to do the same thing here for now. I can certainly see needing to check more than 1 level up for nested Preview method calls. Once I have more feedback from Tanner/Immo, I'll change/keep the logic here.

mavasani reviewed (commented):
Mostly LGTM as well - will let @buyaa-n handle the final sign off for merge. I would certainly recommend doing performance measurements for this analyzer as it seems equally expensive as PlatformCompatAnalyzer.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1674,4 +1674,19 @@
+    <value>An assembly has to opt into preview features before using them.</value>
+  </data>
+  <data name="DetectPreviewFeaturesMessage" xml:space="preserve">
+    <value>Using '{0}' requires opting into preview features. See https://aka.ms/dotnet-warnings/preview-features for more information.</value>
```
Not sure if the link should be dropped in favor of the help link (which will contain the rule doc at some point)

pgovind replied:
Good catch! This is still being discussed. It's minor enough that I'm fine with it going in as is and once the docs team decided which way to go, I'll change the error message

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1674,4 +1674,19 @@
+  <data name="DetectPreviewFeaturesTitle" xml:space="preserve">
+    <value>This API requires opting into preview features</value>
+  </data>
+  <data name="ImplementsEmptyPreviewInterface" xml:space="preserve">
```
Very minor: Consider renaming this to end in "Message".

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1674,4 +1674,19 @@
+  <data name="ImplementsEmptyPreviewInterface" xml:space="preserve">
+    <value>'{0}' implements a preview interface. See https://aka.ms/dotnet-warnings/preview-features for more information.</value>
+  </data>
+  <data name="StaticAndAbstractRequiresPreviewFeatures" xml:space="preserve">
+    <value>Using both static and abstract modifiers requires opting into preview features. See https://aka.ms/dotnet-warnings/preview-features for more information.</value>
+  </data>
```
Consider locking "static" and "abstract" from localization.

pgovind replied:
Are single quotes the way to do this? I don't see many other examples in the resx file. Changed it to 'static' and 'abstract' now.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:
```diff
@@ -0,0 +1,304 @@
+        private static bool TypeSymbolHasAttribute(ITypeSymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, ConcurrentDictionary<ISymbol, PreviewFeatureUsageType> requiresPreviewFeaturesSymbolsToUsageType)
+        {
+            ImmutableArray<INamedTypeSymbol> interfaces = symbol.Interfaces;
+            foreach (INamedTypeSymbol? anInterface in interfaces)
```
```suggestion
            foreach (INamedTypeSymbol anInterface in interfaces)
```

Youssef1313 commented on src/Utilities/Compiler/WellKnownTypeNames.cs:
```diff
@@ -301,6 +302,7 @@ internal static class WellKnownTypeNames
         public const string SystemRuntimeSerializationStreamingContext = "System.Runtime.Serialization.StreamingContext";
         public const string SystemRuntimeVersioningSupportedOSPlatformAttribute = "System.Runtime.Versioning.SupportedOSPlatformAttribute";
         public const string SystemRuntimeVersioningUnsupportedOSPlatformAttribute = "System.Runtime.Versioning.UnsupportedOSPlatformAttribute";
+        public const string SystemRuntimeVersioningRequiresPreviewFeaturesAttribute = "System.Runtime.Versioning.RequiresPreviewFeaturesAttribute";
```
Very minor: Consider keeping them ordered alphabetically.

Youssef1313 commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureUnitTests.cs:
```diff
@@ -0,0 +1,851 @@
+        [Fact]
+        public async Task TestPreviewMethodUnaryOperator()
+        {
+            var csInput = @" 
+using System.Runtime.Versioning; using System;
+namespace Preview_Feature_Scratch
+{" +
+@"
+    public class Program
+    {
+        static void Main(string[] args)
+        {
+            var a = new Fraction();
+            var b = [|+a|];
```
I'm not sure why this is split into two strings being concatenated?

pgovind replied:
There used to be another string after the `+`, but it's since been removed, so the `+` is now unnecessary. It's just painful to remove the `+` for every single test case right now, so I just left it in (since it's just unit tests)

Youssef1313 commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureUnitTests.cs:
```diff
@@ -0,0 +1,851 @@
+        }
+
+        // TODO: The following test cannot be activated yet because it requires preview roslyn bits
+        //        [Fact]
+        //        public async Task TestPreviewLanguageFeatures()
+        //        {
+        //            var csInput = @" 
+        //using System.Runtime.Versioning; using System;
+        //namespace Preview_Feature_Scratch
+        //{"+
+        //@"
+
+        //    class Program : IProgram
```
I'd uncomment it and use `[Fact(Skip = "Reason")]`. That way it's easier to track enabling the test.

pgovind commented:
Thank you for the review everyone! @mavasani @buyaa-n @Youssef1313 @AraHaan
